lsp/ts_ls.lua

```diff
@@ -78,4 +81,41 @@ return {
     'typescript.tsx',
   },
   root_markers = { 'tsconfig.json', 'jsconfig.json', 'package.json', '.git' },
+  on_attach = function(client)
+    -- ts_ls provides code actions that have a prefix `source.` that apply to the whole file. These are not discoverable
+    -- via the normal `vim.lsp.buf.code_action()` so add a command that allows access to them.
+    local function get_source_actions()
+      local codeActionKinds = client.server_capabilities.codeActionProvider.codeActionKinds
+      return vim.tbl_filter(function(action)
+        return vim.startswith(action, 'source.')
+      end, codeActionKinds)
+    end
+
+    local function complete(arg)
+      return vim.tbl_filter(function(action)
+        return vim.startswith(action, arg)
+      end, get_source_actions())
+    end
+
+    vim.api.nvim_buf_create_user_command(0, 'LspTypescriptSourceAction', function(opts)
+      local params
+      if #opts.fargs == 0 then
+        -- ask the server what available actions there are
+        params = {
+          context = {
+            only = get_source_actions(),
+          },
+        }
+      else
+        -- with a specific action, apply it immediately
```

Comment:
This is an extra interface. Configs generally shouldn't invent new interfaces. The `code_actions()` menu is already sufficient. If someone wants to map something, they can just call `code_action()` with the right parameters.

Reply:
I've removed the argument functionality.